Fix #7830, Added a custom lint check to ensure that there is an empty line between imports and fileoverview.

donosco99 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
fix #7830 .
Added custom lint check to ensure that there is an empty line between imports and fileoverview
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python -m scripts.pre_commit_linter` and `python -m scripts.run_frontend_tests`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [x] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

donosco99 committed:
added a custom lint check to check for space between file overview and imports

donosco99 committed:
Added custom lint check to ensure that there is an empty line between imports and fileoverview

donosco99 committed:
added a custom lint check to ensure a space between file overview and imports

donosco99 commented:
@kevinlee12 , PTAL

donosco99 commented:
@kevinlee12, Do you have any idea why the travis-ci test is failing? I am not able to deduce anything from the logs could pls assist?
Thanks

kevinlee12 commented on scripts/pylint_extensions.py:
```diff
@@ -1264,6 +1265,70 @@ def process_module(self, node):
+        'C0020': (
+            'No space used between file overview and import.',
+            'no space used between file overview and import',
+            'please provide a space between file overview and imports.'
```
Line 1276 and 1277 is repeated, remove the second one.

donosco99 replied:
done

Hudda replied:
replace _space_ with _newline_

kevinlee12 commented on scripts/pylint_extensions.py:
```diff
@@ -1264,6 +1265,70 @@ def process_module(self, node):
+        'C0021': (
+            'Single space should be provided above imports.',
+            'single space should be provided above imports',
+            'please provide a single new line above imports.'
```
Ditto

donosco99 replied:
done

Hudda replied:
Ditto

kevinlee12 reviewed (commented):
thanks @donosco98, I've left some comments.

@Hudda can you also look at this?

kevinlee12 commented:
re failing e2e test, it's most likely a result of a flaky tests, I've restarted them.

donosco99 committed:
merge with master

donosco99 committed:
resolve merge conflicts

donosco99 committed:
added lint check to ensure that there is space between imports and fileoverview

donosco99 committed:
Merge branch 'space_below_fileoverview_checker' of https://github.com/donosco98/oppia into space_below_fileoverview_checker

donosco99 commented:
@kevinlee12 , PTAL!

donosco99 committed:
Merge branch 'develop' into space_below_fileoverview_checker

donosco99 committed:
added lint check to ensure that there is an empty line between file overview and imports

donosco99 commented:
@kevinlee12 , @Hudda PTAL!

kevinlee12 commented on scripts/pylint_extensions_test.py:
```diff
@@ -1973,3 +1974,121 @@ def test_well_formed_multi_line_docstring(self):
             temp_file.close()
+
+
+class SpaceBelowFileOverviewCheckerTests(unittest.TestCase):
```
Also add a test where there's a unicode character.

donosco99 replied:
done

donosco98 replied:
I've also added some tests with cases having appropriate space between file overview and imports

kevinlee12 reviewed (changes requested):
thanks, I have one more comment!

donosco99 committed:
Merge branch 'develop' into space_below_fileoverview_checker

donosco99 committed:
added lint check to ensure that there is an empty line between file overview and imports

donosco99 commented:
@kevinlee12, PTAL!

donosco99 commented:
@kevinlee12,circle-ci lint tests have timed out for some reason. Any idea why?

kevinlee12 commented:
It’s a known issue, I’ve restarted the linter.

donosco99 commented:
@kevinlee12, the lint tests are still timing out. Should I merge with master and push again. Other than that, should I make any more changes?

kevinlee12 commented:
Hmm, try running it locally and see if it causes the same issue. Since this PR affects the linter, it might indicate a bug in the code.

donosco99 committed:
added lint check to ensure that there is an empty line between file overview and imports

donosco99 committed:
added lint check to ensure that there is an empty line between file overview and imports

donosco98 commented:
@kevinlee12, I have tested linting all files locally, all checks pass and I see no problems. Also, I noticed that the tests are timing out at different files each time we run it on circle-ci.
Do you have any idea why it is failing on circle-ci, Is there some problem with the branch?
Should I make a new PR?

donosco99 commented:
@kevinlee12, In almost all the python files there is no space between file-overview and imports.
So while running the linter locally, the lint check which I added in this PR is not passing.
![import](https://user-images.githubusercontent.com/31371571/72667676-59fa8980-3a44-11ea-9882-4ecee06fdc7f.png)
I think this is why the circle-ci tests are failing. Should we add a new line between file overview and imports in the existing files that don't pass this lint check? What is your opinion on this?

donosco99 commented:
Also @kevinlee12, previously while adding new lint checks, where the old files required to be modified in order for the recent checks to pass?

kevinlee12 commented:
Try fixing the issues before pushing and see if that changes things. You might also have to rewrite the lint checks if they are taking too long too. CircleCI is more strict in terms of checks taking longer than they should be.

donosco99 commented:
@kevinlee12 ,do you mean to say that we add a new line after file overview in all the python files?(this will be a quite a lot of files)
Also, I will try to make the lint check a bit more faster.

kevinlee12 commented:
Yes, all lint errors must be fixed before we can merge this. Before you do that, I'd focus on making this lint check faster.

donosco99 committed:
Merge branch 'develop' into space_below_fileoverview_checker

donosco99 committed:
added lint check to ensure that there is a new line between file overview and imports. Also modified existing python files to comply with the lint check

donosco99 committed:
Merge branch 'develop' into space_below_fileoverview_checker

donosco99 committed:
added lint check to ensure that there is a new line between file overview and imports. Also modified existing python files to comply with the lint check

donosco99 commented:
@kevinlee12! PTAL!, locally all the checks including the lint checks are working fine. Also, I've modified the code to make it a bit more efficient. 
Also If you feel the code could be improved further pls suggest. Thanks.

kevinlee12 reviewed (approved):
lgtm from my end.

@Hudda can you do a second pass? thanks!

kevintab95 reviewed (approved):
LGTM, thanks @donosco98!

Hudda reviewed (approved):
LGTM, thanks @donosco98

kevinlee12 commented:
@seanlip since this is a mechanical change, can we merge without all of code owners?

seanlip commented on scripts/pylint_extensions.py:
```diff
@@ -1383,6 +1384,60 @@ def process_module(self, node):
+                continue
+            triple_quote_counter += line.count(multi_line_indicator)
+            previous_line = file_content[line_num - 1].strip()
+            if (line.startswith(b'import') or line.startswith(b'from')):
```
The logic here seems wrong. Once you've encountered the fileoverview line closing quote and are on the next line, just check whether that next line is empty. The check should be independent of the contents of that line.

You can also check for the longest consecutive run of empty or whitespace-only lines starting at that point and verify that the count of lines in that run is equal to 1.

Hopefully this helps to simplify the logic.

Also, make sure to consider the case in which the file doesn't have a fileoverview comment, and the first triple-quote comment encountered is in a class/method docstring. We should detect that the fileoverview is missing. I'm not sure if there's an existing pylint check in place for this -- if so, you can just mention that here so that it's clear what assumptions you are working with.

donosco99 replied:
> We should detect that the fileoverview is missing

@seanlip, there is an inbuilt checker by Pylint for missing file overview. But if I add a test case having no fileoverview it can't detect it.
![docstri](https://user-images.githubusercontent.com/31371571/72880157-5573f080-3d24-11ea-88a6-676da9ee85e2.png)

So should we do something so that missing fileoverview is detected in test cases?

seanlip replied:
First, see if you can find out why it doesn't detect it (in general, this is always a good practice -- figure out what's wrong before you start modifying things). Based on the resulting analysis, we can figure out what to do next.

donosco99 replied:
> if so, you can just mention that here

Sorry I didn't completely understand what you meant here

seanlip replied:
I meant mentioning it in a comment.

donosco99 replied:
> see if you can find out why it doesn't detect it

I guess in a test case only the linter being tested is activated .
For example in my case only `SpaceBelowFileOverviewChecker` is activated
`     self.checker_test_object.CHECKER_CLASS = (
            pylint_extensions.SpaceBelowFileOverviewChecker)
`
Pls correct me If I'm missing something

donosco99 replied:
@seanlip, So should we still add the test case having no file overview?

seanlip replied:
> I guess in a test case only the linter being tested is activated .

What checks have you done to verify this statement? Please describe how you came to that conclusion (it needs to be more than a guess, it should be supported by evidence).

donosco99 replied:
@seanlip, I have verified the above by changing the `self.checker_test_object.CHECKER_CLASS` at the beginning of the testing class.
Example: in the case when I register `SpaceBelowFileOverviewChecker` as `CHECKER_CLASS` and give excessive lines between  function definitions(which is detected by `ExcessiveEmptyLinesChecker` class) , the excess lines are not detected.
But when I register `ExcessiveEmptyLinesChecker` as the `CHECKER_CLASS`, It detects the excess lines.

seanlip replied:
That's good. (Also, in the future, please include the analysis as a matter of course when stating any conclusions.)

There is only one check left to do: can you verify that pylint itself catches missing fileoverviews when the full linting process is done? If so then you can go ahead with what I said above: leave a comment explaining the assumption you're making about the existence of such a fileoverview (or, to be more rigorous, you can check that the corresponding rule isn't excluded in .pylintrc), and you don't need to add a test here with missing fileoverview since the assumption underlying everything here is that the fileoverview exists.

donosco99 replied:
@seanlip, yes I verified that Pylint does throw a message if there is no file overview.
Although the modified code works as expected even if there is no file-overview.

seanlip replied:
Hm ... when you say "the modified code works as expected even if there is no file-overview", what do you mean exactly? I.e. what is the "expected behaviour" if there's no file overview?

(For me this case is "undefined" since we are assuming, by virtue of the other pylint check, that a fileoverview will always be present.)

donosco99 replied:
I added an extra check for not mistaking the triple quote in Args of method/class as file overview.(As this happens only when there is no file overview I'll remove this check)

donosco99 replied:
Sorry for the confusion

seanlip commented:
@kevinlee12 I agree that this does not need review from the rest of the codeowners. Thanks for flagging!

donosco99 committed:
added lint check to ensure that there is a new line between file overview and imports. Also modified existing python files to comply with the lint check

donosco99 committed:
added lint check to ensure that there is a new line between file overview and imports. Also modified existing python files to comply with the lint check

donosco99 committed:
added lint check to ensure that there is a new line between file overview and imports. Also modified existing python files to comply with the lint check

donosco99 committed:
added lint check to ensure that there is a new line between file overview and imports. Also modified existing python files to comply with the lint check

donosco99 committed:
added lint check to ensure that there is a new line between file overview and imports. Also modified existing python files to comply with the lint check

donosco99 committed:
added lint check to ensure that there is a new line between file overview and imports. Also modified existing python files to comply with the lint check

donosco99 commented:
@seanlip, PTAL! (pls note that I've used the variable `imports_detected` in order for the check to work even when there is no file-overview)
Edit: I've removed the extra check(`imports_detected`) after the previous discussion

donosco99 committed:
added lint check to ensure that there is a new line between file overview and imports. Also modified existing python files to comply with the lint check

seanlip commented on scripts/pylint_extensions.py:
```diff
@@ -1415,25 +1415,27 @@ def process_module(self, node):
+
+            if line.endswith(b'"""') and triple_quote_counter == 2:
+                temp_line_num = line_num
+                while file_content[temp_line_num + 1] == b'\n':
```
This is careless coding. Should check that temp_line_num + 1 actually exists in the list before referencing it.

Also I suggest not doing all this logic/manipulation in the for loop -- you don't need to. Just store the variables and break, then do the next part in a separate block. 

Also, name `temp_line_num` something more clear semanitcally -- e.g. closing_line_index_of_fileoverview or similar. If you want the code to be maintained, it is important to make all of it easily understandable by fellow developers.

donosco99 replied:
done.
I'm sorry for these mistakes, I'm still learning about good coding practices in Python. I'll try to avoid these mistakes in the future.

seanlip commented on scripts/pylint_extensions.py:
```diff
@@ -1415,25 +1415,27 @@ def process_module(self, node):
-                if ((file_content[line_num - 1] == b'\n' and
-                     file_content[line_num - 2] == b'\n') and
-                        (triple_quote_counter == 2)):
+                if empty_line_counter == 0:
```
elif

donosco99 replied:
done

seanlip commented on scripts/pylint_extensions_test.py:
```diff
@@ -1985,6 +1985,7 @@ def setUp(self):
             pylint_extensions.SpaceBelowFileOverviewChecker)
         self.checker_test_object.setup_method()
 
+    # The checks assume that all files have a file overview.
```
You need to explain why we are allowed to make this assumption.

I also suggest moving this to the class docstring. I.e. write a proper class docstring explaining what the class this, and then add this as a "Note: " within the docstring. Otherwise, this comment isn't really discoverable, and it seems like it only relates to the method just below it.

donosco99 replied:
done

seanlip commented:
@donosco98 Please also update your PR title to follow the standard conventions for this repository. Instructions are provided in the checkboxes and in the "getting started" guide. Make sure to use proper capitalization and punctuation for the "fix XXX" part.

donosco99 committed:
added lint check to ensure that there is a new line between file overview and imports. Also modified existing python files to comply with the lint check

donosco99 committed:
Added lint check to ensure that there is a new line between file overview and imports. Also modified existing python files to comply with the lint check.

donosco99 committed:
Added lint check to ensure that there is a new line between file overview and imports. Also modified existing python files to comply with the lint check.

donosco99 commented:
@seanlip , PTAL!

seanlip commented:
Hi @donosco98 -- btw, I forgot to mention one other thing: please make each of your commits more descriptive. When you give them all the same description, it's hard for the reviewer to figure out which one to start reviewing from when they do a subsequent pass. Thanks!

seanlip commented on scripts/pylint_extensions.py:
```diff
@@ -1385,7 +1385,11 @@ def process_module(self, node):
 
 class SpaceBelowFileOverviewChecker(checkers.BaseChecker):
-    """Checks if there is an empty line between file overview and imports."""
+    """Checks if there is an empty line between file overview and imports.
```
This docstring should not refer to imports. Perhaps something like "the next non-empty line", or you can just say "a single empty line below the fileoverview docstring".

donosco99 replied:
done

seanlip commented on scripts/pylint_extensions.py:
```diff
@@ -1385,7 +1385,11 @@ def process_module(self, node):
 class SpaceBelowFileOverviewChecker(checkers.BaseChecker):
-    """Checks if there is an empty line between file overview and imports."""
+    """Checks if there is an empty line between file overview and imports.
+       Note: The check assumes that all files have a file overview, this
```
Indentation is wrong. Deindent this and the following lines by 3.

Also change the comma to a period and start a new sentence with "This".

donosco99 replied:
done

seanlip commented on scripts/pylint_extensions.py:
```diff
@@ -1385,7 +1385,11 @@ def process_module(self, node):
+    """Checks if there is an empty line between file overview and imports.
+       Note: The check assumes that all files have a file overview, this
+       assumption is justified because Pylint has an inbuilt check for missing
+       file overview.
```
Specify, in parentheses, what the name of this inbuilt check is. Make things as easy for the reader of the code as possible, since you already have the context.

donosco99 replied:
done

seanlip commented on scripts/pylint_extensions.py:
```diff
@@ -1424,20 +1428,23 @@ def process_module(self, node):
+                break
+
+        empty_line_check_index = closing_line_index_of_fileoverview
+        if empty_line_check_index < file_length:
```
Still careless -- I think there's an off-by-one error.

I suggest that you add a backend test where there's no newline after the fileoverview string to verify that the behaviour in that case is correct.

donosco99 replied:
Yeah @seanlip, I tested the behavior by making file overview the last line.
hence we require `empty_line_check_index` to be less than `file_length-1`.

donosco99 replied:
done

donosco99 committed:
Added lint check to ensure that there is a new line between file overview and imports. Also modified existing python files to comply with the lint check.
 - Adress review comments
 - Added descriptive docstring for the checker class
 - Added a test case for checking the edge case of file overview being in the      last line

donosco99 committed:
Added lint check to ensure that there is a new line between file overview and imports. Also modified existing python files to comply with the lint check.
 - Adress review comments
 - Added descriptive docstring for the checker class
 - Added a test case for checking the edge case of file overview being in the last line

donosco99 commented:
@seanlip, PTAL!

seanlip commented:
@donosco98, I am taking a pass now, but I would like to remind you of this comment, which you have still not followed: https://github.com/oppia/oppia/pull/8409#issuecomment-577536674. Thanks.

seanlip commented on scripts/pylint_extensions.py:
```diff
@@ -1384,32 +1384,32 @@ def process_module(self, node):
+            'No empty line used below the fileoverview docstring.',
             'no-empty-line-provided-below-fileoverview',
-            'please provide a line between file overview and imports.'
+            'please provide a line below the fileoverview.'
```
a line --> an empty line

donosco98 replied:
done

donosco99 replied:
done

seanlip commented on scripts/pylint_extensions.py:
```diff
@@ -1384,32 +1384,32 @@ def process_module(self, node):
+    """Checks if there is a single empty line below the fileoverview docstring.
+    Note: The check assumes that all files have a file overview. This
+    assumption is justified because Pylint has an inbuilt check for missing
+    file overview.( Name of this check is missing-docstring.)
```
There should be a space before '(' and no space after that character. Also, "Name" should be "The name".

You could alternatively just say: "... because Pylint has an inbuilt check (missing-docstring) for missing file overviews." and drop the last sentence.

donosco99 replied:
done

seanlip reviewed (approved):
Almost there! Just two more comments, otherwise LGTM.

As a reminder, in your subsequent commits, please use more specific commit descriptions describing what's in that commit.

donosco99 committed:
Merge branch 'develop' into space_below_fileoverview_checker

donosco99 committed:
address review comments
 -modified the docstrings
 -modified the error messages

donosco99 committed:
Merge branch 'develop' into space_below_fileoverview_checker

donosco99 committed:
fixed a typo in class docstring

seanlip commented:
Thanks! LGTM, no further comments. Feel free to send a reply here once tests pass and I'll merge this.

donosco99 commented:
@seanlip, The checks have passed you can merge the PR.

seanlip commented:
Thanks! Done :)
